src/rustllvm/ArchiveWrapper.cpp

```diff
@@ -129,27 +133,40 @@ LLVMRustArchiveIteratorNew(LLVMRustArchiveRef ra) {
 
 extern "C" LLVMRustArchiveChildConstRef
 LLVMRustArchiveIteratorNext(LLVMRustArchiveIteratorRef rai) {
+    // Advancing the iterator validates the next child, and this can
+    // uncover an error. LLVM requires that we check all Errors,
+    // so we only advance the iterator if we actually need to fetch
+    // the next child.
+    // This means we must not advance the iterator in the *first* call,
+    // but instead advance it *before* fetching the child in all later calls.
+    if (!rai->first) {
+        ++rai->cur;
 #if LLVM_VERSION_GE(3, 9)
-    if (rai->err) {
-        LLVMRustSetLastError(toString(std::move(rai->err)).c_str());
-        return NULL;
-    }
+        if (rai->err) {
+            LLVMRustSetLastError(toString(std::move(rai->err)).c_str());
+            return nullptr;
+        }
 #endif
-    if (rai->cur == rai->end)
-        return NULL;
+    }
+
+    if (rai->cur == rai->end) {
```

Comment:
I think this should happen first to prevent a segfault from calling `next` too much, right?